contracts/AINFT721.sol

```diff
@@ -0,0 +1,304 @@
+    string private baseURI;
+    mapping(bytes32 => MetadataContainer) private _metadataStorage; // keccak256(bytes(tokenId, <DELIMETER>, version)) => MetadataContainer
+    mapping(uint256 => uint256) private _tokenURICurrentVersion; // tokenId => tokenURIVersion
+
+    constructor(string memory name_, string memory symbol_, bool isCloned_, address originNFT_) ERC721(name_, symbol_) {
+        if (isCloned_) {
+            //FIXME: need to use tx.origin?
+            _grantRole(DEFAULT_ADMIN_ROLE, tx.origin);
+            _grantRole(PAUSER_ROLE, tx.origin);
+            _grantRole(MINTER_ROLE, tx.origin);
+
+            
+        } else {
+            //FIXME: the originNFT contract address should not be ROLED
+            _grantRole(DEFAULT_ADMIN_ROLE, tx.origin);
+            _grantRole(PAUSER_ROLE, tx.origin);
+            _grantRole(MINTER_ROLE, tx.origin);
+        }
+        ORIGIN_NFT = IERC721(originNFT_);
+        IS_CLONED = isCloned_;
+        require((address(ORIGIN_NFT) == address(0) && !IS_CLONED) ||
+                (address(ORIGIN_NFT) != address(0) && IS_CLONED), 
+                "If AINFT721 is created first time, the originNFT_ should not be SET. \
+                \n If AINFT721 is cloned by something, the originNFT_ should be SET.");
+    }
+
+    modifier Cloned () {
+        require(IS_CLONED, "Only cloned contract can execute.");
+        _;
+    }
+    
+    function mintFromOriginInstance(
+        uint256 tokenId_
+    ) public Cloned {
+        require(!_exists(tokenId_), "The tokenId_ is already minted or cloned");
+        address originOwner = ORIGIN_NFT.ownerOf(tokenId_);
+        require(_msgSender() == originOwner, "The sender should be the holder of origin NFT.");
+        _safeMint(_msgSender(), tokenId_);
+
+    }
+    
+    function mintFromOriginInstanceOnBehalf(
+        uint256[] calldata tokenIds_,
+        address[] calldata recipients_
+    ) public Cloned onlyRole(MINTER_ROLE) {
+        for (uint i = 0; i < tokenIds_.length; i++) {
+            require(!_exists(tokenIds_[i]), "The tokenId_ is already minted or cloned");
+            require(recipients_[i] == ORIGIN_NFT.ownerOf(tokenIds_[i]), "The sender should be the holder of origin NFT.");
+            _safeMint(recipients_[i], tokenIds_[i]);
+        }
+    }
+
+    function setPaymentContract(address paymentPlugin_) public onlyRole(DEFAULT_ADMIN_ROLE) {
+        require(_msgSender() == tx.origin && _msgSender() != address(0), "Only EOA can set payment contract.");
+        PAYMENT_PLUGIN = paymentPlugin_;
+    }
+
+    /**
+     * @dev See {IERC4906}.
+     */
+    function supportsInterface(
+        bytes4 interfaceId
+    )
+        public
+        view
+        override(
+            AccessControl,
+            ERC721Enumerable,
+            ERC721,
+            IERC165
+        )
+        returns (bool)
+    {
+        // IERC4906 interface added
+        return interfaceId == bytes4(0x49064906) || super.supportsInterface(interfaceId);
```

Comment:
FYI (might help, but I don't know yet): https://docs.openzeppelin.com/contracts/3.x/api/introspection#IERC1820Registry-interfaceHash-string-

Reply:
I agree that we'd better add our own interfaceId. Maybe we have to think deeply about it further, I leave it as future work.

Reply:
But in this quarter plan, the `marketplace` should be searchable all AINFTs. without own interfaceId or without AINFTFactory's managed list, it would be quite tricky.

Reply:
Regarding to marketplace, it would be necessary to distinguish AINFT from other ERC721 NFTs, so setting our own interfaceId should be ASAP!

Reply:
It would be discussed further in #5 . Thanks!